### ISSUE TITLE
Cell barcode correction is wrong

### ISSUE THREAD

**Issue by ambrosejcarr** [HumanCellAtlas/sctools#17]
Right now, we do this to identify a cell barcode: 
```
    def extract_cell_barcode(self, record, cb):
        seq_tag, qual_tag = extract_barcode(record, cb)
        try:
            corrected_cb = self._error_mapping.get_corrected_barcode(seq_tag[1])
        except KeyError:
            corrected_cb = seq_tag[1]  # couldn't find the right barcode!
        return seq_tag, qual_tag, ('CB', corrected_cb, 'Z')
```

This is bad, because it fails to separate categories of reads that should not be combined: 
Tags on read with correctible cell barcode: `CR`, `CB`, `CR` different from `CB`
Tags on read with perfect cell barcode: `CR`, `CB`, `CR` same as `CB`
Tags on read with nonsense cell barcode: `CR`, `CB`, `CR` same as `CB`

If instead we only store `CB` when we identify a perfect or correctible barcode, these three categories of reads would become separable. 

However, it would break some existing functionality. 

1. `SplitBamByCellBarcode`: currently we assign reads to file by the `CB` tag. This change would break this functionality. To fix this, we could default to `CR` in a `try-except` expression when `CB` is not found. 

2. `Metrics`: cell metrics aggregate information by cell barcode. reads without cell barcode would be excluded. This may be the optimal solution, since they definitively do not represent cells. Conversely, it would be important to know how many observations of genes occur in cells vs non-cells, so the number of reads aligning to genes but containing non-cell barcodes should be tracked, and should be comparable to the number of reads that do associate with known cell barcodes. 

3. Picard UmiAwareMarkDuplicates: I assume it will break because it's picard. Technically it doesn't use the CB tag, so we can probably remove it.

**Comment by ambrosejcarr** [comment]
Looking for some feedback here on this problem + implementation proposal.

**Comment by jsotobroad** [comment]
So I guess one worry I have for this is that this no one tag that someone can look at to know what is the "best" cell barcode for all reads in a bam which I thought was important for some reason (I could be totally off here.

My one question is what is the impact of having these barcodes that were not correctible with our current scheme still have a `CB` value.

2.  I think you could still go through the reads with "None" `CB` values and calculate some relevant metrics if you wanted to after you made this change
 
3. it actually currently wouldn't affect anything because it doesn't even use the tag like you pointed out.  Even when it is changed to use the CB tag to mark duplicates it should just ignore it if one is not present.

**Comment by ambrosejcarr** [comment]
Yea I thought about this. Unfortunately `pysam` can't write "null" tag values. We'd need to make up a value that means that. Like a string `null` or `None`. 

when you try to call `record.set_tag("CB", "Z", ["" | None])` it deletes the tag if the third argument evaluates to `False`.

**Comment by ambrosejcarr** [comment]
I actually like the idea that they don't have a `CB` value, since by doing it the way I've proposed we're more or less asserting that they didn't get a cell barcode. That's in line with them having some gibberish sequence there as opposed to a barcode that corresponds to, or can be corrected to come from, the whitelist.

**Comment by ambrosejcarr** [comment]
Another advantage of allowing CB to default to CR when CB is not present is that it will help us to identify failure modes, e.g. when thousands of reads have a sequencing primer where the cell barcode should be. 

Admittedly, this would also be possible in the current scheme, where CB does not always reflect a real barcode, but I don't think that approach follows the spirit of what CB is supposed to represent.

**Comment by jsotobroad** [comment]
Gotcha, if CB is supposed to represent a white listed cell barcode then I agree with it just not being present in these reads with gibberish cell barcodes and only have CR values.  With that being said is that what the standard is?  Are there people outside of our group that we should be asking their opinion on this?

**Comment by ambrosejcarr** [comment]
Standard is still in flux but here is what it says right now: https://github.com/samtools/hts-specs/pull/239

**Comment by ambrosejcarr** [comment]
Thanks @jsotobroad. @TimothyTickle @jishuxu  Would either of you like to weigh in on this before I implement the change?

**Comment by jishuxu** [comment]
if `CR` can not be corrected, what you will do with these reads?

**Comment by ambrosejcarr** [comment]
Eventually they will be excluded from the count matrix. Otherwise they will be fully processed.

**Comment by jishuxu** [comment]
if it is that case, why don't you just flush these `un-correctable barcodes` out and dump these fastq reads into a `undetermined.fastq` file? Cound it be easier at the end?

**Comment by ambrosejcarr** [comment]
That's a reasonable solution, and one that we've considered. 

We avoided that approach because when possible we'd like to avoid making decisions about reads, and maximize the amount of information associated with each one. If we flushed them into undetermined, they wouldn't be subjected to a lot of downstream analysis that might yield information on _why_ the reads failed this step. For example, as implemented, reads with no cell barcode will still be aligned, deduplicated, and included in metric calculations for per-gene analysis. 

This can help understand what's going on with these reads, and I hope will empower experimenters to learn about how they can adjust their experiments to improve yield on future runs.

**Comment by jishuxu** [comment]
okay.  just curious, what's the `%` of these `broken` barcodes are?   as for design, either way is fine.

**Comment by ambrosejcarr** [comment]
That's a great question. In InDrop data it was normally around 5-10%. I'll check in 10x. :)

**Comment by ambrosejcarr** [comment]
closed by #18
